src/Data/Bin/Bin0.agda

```diff
@@ -0,0 +1,655 @@
+  in
+  begin fromℕ (toℕ a)   ≡⟨ refl ⟩
+        b                 ≡⟨ toℕ-injective toℕ-b≡toℕ-a ⟩
+        a
+  ∎
+
+------------------------------------------------------------------------------
+-- Summary:
+-- 1)  toℕ : Bin → ℕ  is an isomorphism by _+_,
+-- 2)  fromℕ            is a homomorphisms by _+_ mutually inverse with toℕ.
+------------------------------------------------------------------------------
+
+fromℕ∘pred :  (n : ℕ) → fromℕ (pred' n) ≡ pred (fromℕ n)
+fromℕ∘pred n =
+  let (x , toℕ-x≡n) = ext-fromℕ n
+  in
+  begin
+    fromℕ (pred' n)          ≡⟨ cong (fromℕ ∘ pred') (sym (toℕ∘fromℕ n)) ⟩
+    fromℕ (pred' (toℕ x))   ≡⟨ cong fromℕ (sym (toℕ∘pred x)) ⟩
+    fromℕ (toℕ (pred x))    ≡⟨ fromℕ∘toℕ (pred x) ⟩
+    pred x                    ≡⟨ refl ⟩
+    pred (fromℕ n)
+  ∎
+
+
+------------------------------------------------------------------------------
+-- Commutativity and associativity for _+_ are proved by using the isomorphism
+-- to ℕ.
+------------------------------------------------------------------------------
+
++-comm :  Commutative _+_
++-comm a b = begin
+  a + b                        ≡⟨ sym (fromℕ∘toℕ (a + b)) ⟩
+  fromℕ (toℕ (a + b))        ≡⟨ cong fromℕ (toℕ+homo a b) ⟩
+  fromℕ (toℕ a +' toℕ b)    ≡⟨ cong fromℕ (+'-comm (toℕ a) (toℕ b)) ⟩
+  fromℕ (toℕ b +' toℕ a)    ≡⟨ cong fromℕ (sym (toℕ+homo b a)) ⟩
+  fromℕ (toℕ (b + a))        ≡⟨ fromℕ∘toℕ (b + a) ⟩
+  b + a                        ∎
+
++-assoc :  Associative _+_
++-assoc a b c = begin
+  (a + b) + c                    ≡⟨ sym (fromℕ∘toℕ ((a + b) + c)) ⟩
+  fromℕ (toℕ ((a + b) + c))    ≡⟨ cong fromℕ (toℕ+homo (a + b) c) ⟩
+  fromℕ (toℕ (a + b) +' cN)    ≡⟨ cong (fromℕ ∘ (_+' cN)) (toℕ+homo a b) ⟩
+  fromℕ ((aN +' bN) +' cN)      ≡⟨ cong fromℕ (+'-assoc aN bN cN) ⟩
+  fromℕ (aN +' (bN +' cN))      ≡⟨ cong (fromℕ ∘ (aN +'_))
+                                                  (sym (toℕ+homo b c)) ⟩
+  fromℕ (aN +' toℕ (b + c))    ≡⟨ cong fromℕ (sym (toℕ+homo a (b + c))) ⟩
+  fromℕ (toℕ (a + (b + c)))    ≡⟨ fromℕ∘toℕ (a + (b + c)) ⟩
+  (a + (b + c))                  ∎
+  where
+  aN = toℕ a;   bN = toℕ b;   cN = toℕ c
+
+
++-lCancel :  ∀ x y z → x + y ≡ x + z → y ≡ z      -- the proof is via toℕ
++-lCancel x y z x+y≡x+z = begin
+  y          ≡⟨ sym (fromℕ∘toℕ y) ⟩
+  fromℕ m   ≡⟨ cong fromℕ m≡n ⟩
+  fromℕ n   ≡⟨ fromℕ∘toℕ z ⟩
+  z          ∎
+  where
+  k = toℕ x;  m = toℕ y;  n = toℕ z
+
+  eq = begin
+    k +' m          ≡⟨ sym (toℕ+homo x y) ⟩
+    toℕ (x + y)    ≡⟨ cong toℕ x+y≡x+z ⟩
+    toℕ (x + z)    ≡⟨ toℕ+homo x z ⟩
+    k +' n          ∎
+
+  m≡n = begin
+    m                ≡⟨ sym (NatP.m+n∸n≡m m k)⟩
+    (m +' k) ∸' k    ≡⟨ cong (_∸' k) (+'-comm m k) ⟩
+    (k +' m) ∸' k    ≡⟨ cong (_∸' k) eq ⟩
+    (k +' n) ∸' k    ≡⟨ cong (_∸' k) (+'-comm k n) ⟩
+    (n +' k) ∸' k    ≡⟨ NatP.m+n∸n≡m n k ⟩
+    n                ∎
+
++-rCancel :  ∀ x y z → y + x ≡ z + x → y ≡ z
++-rCancel x y z y+x≡z+x =  +-lCancel x y z x+y≡x+z
```

Comment:
Same here with `+-cancelʳ-≡ : RightCancellative _≡_ _+_`.

Reply:
See also: https://agda.github.io/agda-stdlib/Algebra.FunctionProperties.Consequences.html#1079
to refactor this function.